### ISSUE TITLE
FTBFS on RHEL-7: if_perl.xs:57:20: fatal error: EXTERN.h: No such file or directory

### ISSUE THREAD

**Issue by mcepl** [vim/vim#2091]
I build the latest releases of vim almost daily for my [COPR](https://copr.fedorainfracloud.org/coprs/mcepl/vim8/). The last build which perfectly succeeded was ``1097``. When I tried now (using the same *.spec file, so I can guarantee, that I have never changed anything in my building steps) with ``1110`` I got this error:
```
gcc -c -I. -DDYNAMIC_PERL_DLL=\"libperl.so\"  -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK  -pthread -I/usr/include/gtk-3.0 -I/usr/include/atk-1.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/pango-1.0 -I/usr/include/gio-unix-2.0/ -I/usr/include/cairo -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/glib-2.0 -I/usr/lib64/glib-2.0/include -I/usr/include/at-spi-2.0 -I/usr/include/dbus-1.0 -I/usr/lib64/dbus-1.0/include -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/pixman-1 -I/usr/include/libpng15 -I/usr/include/libdrm     -O2 -g -pipe -Wall -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches   -m64 -mtune=generic -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1        -o objects/if_perl.o auto/if_perl.c
if_perl.xs:57:20: fatal error: EXTERN.h: No such file or directory
 #include <EXTERN.h>
                    ^
compilation terminated.
make: *** [objects/if_perl.o] Error 1
error: Bad exit status from /var/tmp/rpm-tmp.1RaQ3C (%build)
```
It doesn't make much sense, because ``EXTERN.h`` is present in ``/usr/lib64/perl5/CORE/EXTERN.h`` (as it always have been) from the standard RHEL-7 package ``perl-devel``.

Complete build log is [_build-8.0.1110-1.0.41.el7.log](https://github.com/vim/vim/files/1304759/_build-8.0.1110-1.0.41.el7.log.txt)

**Comment by ramottamado** [comment]
I too experience the same issue. I think this is related to patch `1110`. Seems that line [5723](https://github.com/vim/vim/blob/e8ff56b28ed67bff4ee1dbe7b4bd620c1c04a535/src/auto/configure#L5723) is missing a `'` based on following error I get when running `./configure --enable-perlinterp`

    auto/configure: command substitution: line 5723: unexpected EOF while looking for matching `''
    auto/configure: command substitution: line 5724: syntax error: unexpected end of file

**Comment by nuko8** [comment]
Will this fix the issue for you?:
```diff
diff --git a/src/configure.ac b/src/configure.ac
index 2c24ae400..1183ab44b 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -1009,7 +1009,7 @@ if test "$enable_perlinterp" = "yes" -o "$enable_perlinter<<<<p" = "dynamic"; then
       perlcppflags=`$vi_cv_path_perl -Mlib=$srcdir -MExtUtils::Embed \
                -e 'ccflags;perl_inc;print"\n"' | sed -e 's/-fno[[^ ]]*//' \
                -e 's/-fdebug-prefix-map[[^ ]]*//g' \
-               -e 's/\(-Wp,\)\?-D_FORTIFY_SOURCE=.//g`
+               -e 's/\(-Wp,\)\?-D_FORTIFY_SOURCE=.//g'`
       dnl Remove "-lc", it breaks on FreeBSD when using "-pthread".
       perllibs=`cd $srcdir; $vi_cv_path_perl -MExtUtils::Embed -e 'ldopts' | \
                sed -e '/Warning/d' -e '/Note (probably harmless)/d' \
```
To use the patch, you need some extra work to update `configure`:  Assuming you are at the top of the source tree,
```
$ cd src
$ make distclean
(Apply the patch)
$ make autoconf
$ ./configure ...
$ make
```
NOTE!!! **Do not run GNU autoconf directly, or the build system breaks.**

**Comment by fusion809** [comment]
I get the same errors on Arch Linux. Oddly on Gentoo Vim 8.0.1110 compiles fine.

**Comment by fusion809** [comment]
This similar patch works for me: 

```patch
From 9643a5b74c818a93204441cb3e99a9a59011153a Mon Sep 17 00:00:00 2001
From: BH2017 <fusion809@users.noreply.github.com>
Date: Fri, 15 Sep 2017 12:15:18 +1000
Subject: [PATCH] Update configure.ac

---
 src/configure.ac | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/configure.ac b/src/configure.ac
index 2c24ae4005..1183ab44b6 100644
--- a/src/configure.ac
+++ b/src/configure.ac
@@ -1009,7 +1009,7 @@ if test "$enable_perlinterp" = "yes" -o "$enable_perlinterp" = "dynamic"; then
       perlcppflags=`$vi_cv_path_perl -Mlib=$srcdir -MExtUtils::Embed \
 		-e 'ccflags;perl_inc;print"\n"' | sed -e 's/-fno[[^ ]]*//' \
 		-e 's/-fdebug-prefix-map[[^ ]]*//g' \
-		-e 's/\(-Wp,\)\?-D_FORTIFY_SOURCE=.//g`
+		-e 's/\(-Wp,\)\?-D_FORTIFY_SOURCE=.//g'`
       dnl Remove "-lc", it breaks on FreeBSD when using "-pthread".
       perllibs=`cd $srcdir; $vi_cv_path_perl -MExtUtils::Embed -e 'ldopts' | \
 		sed -e '/Warning/d' -e '/Note (probably harmless)/d' \
```

EDIT: Except I get this error:

```bash
Starting make in the src directory.
If there are problems, cd to the src directory and run make there
cd src && make install
make[1]: Entering directory '/data/OBS/home:fusion809/gvim-gtk2/src/vim-8.0.1110/src'
Makefile:2357: *** multiple target patterns.  Stop.
make[1]: Leaving directory '/data/OBS/home:fusion809/gvim-gtk2/src/vim-8.0.1110/src'
make: *** [Makefile:39: install] Error 2
```

**Comment by mcepl** [comment]
> @nuko8  Will this fix the issue for you?:

Unfortunately, it doesn't. ``make distclean`` fails:
```
+ cd src
+ make distclean
Makefile:304: auto/config.mk: No such file or directory
rm -f auto/config.cache
if test "Xdistclean" != "Xclean" \
	-a "Xdistclean" != "Xdistclean" \
	-a "Xdistclean" != "Xautoconf" \
	-a "Xdistclean" != "Xreconfig"; then \
    GUI_INC_LOC="" GUI_LIB_LOC="" \
	CC="" CPPFLAGS="" CFLAGS="" \
	LDFLAGS=""  srcdir="." \
	./configure    \
	  \
	   \
	   \
	   \
	   \
	  \
	   \
	  \
	   \
	  \
	; \
fi
cd testdir; make -f Makefile clean
make[1]: Entering directory `/home/matej/build/EXTRAS/vim/vim-8.0.1110/src/testdir'
rm -rf *.out *.failed *.res *.rej *.orig opt_test.vim test.log messages test.out X* viminfo tiny.vim small.vim mbyte.vim mzscheme.vim lua.vim test.ok benchmark.out valgrind.*
make[1]: Leaving directory `/home/matej/build/EXTRAS/vim/vim-8.0.1110/src/testdir'
if test -d po; then \
	cd po; make checkclean; \
fi
Makefile:4: make[1]: Entering directory `/home/matej/build/EXTRAS/vim/vim-8.0.1110/src/po'
../auto/config.mk: No such file or directory
make[1]: *** No rule to make target `../auto/config.mk'.  Stop.
make[1]: Leaving directory `/home/matej/build/EXTRAS/vim/vim-8.0.1110/src/po'
make: *** [testclean] Error 2
error: Bad exit status from /var/tmp/rpm-tmp.Gqx2fT (%prep)
    Bad exit status from /var/tmp/rpm-tmp.Gqx2fT (%prep)
```

**Comment by chrisbra** [comment]
I think I might have caused this, sorry for this.
Hm, I am not sure the distclean is needed.  However it is important to `make autoconf`

**Comment by k-takata** [comment]
If the make command fails because of missing `auto/config.mk`, try `cp config.mk.dist auto/config.mk`.

**Comment by nuko8** [comment]
Yeah, @chrisbra is right.  That `make distclean` is rather a sort of precaution than essential.  So its failure shouldn't matter.

What's most important is to run `make autoconf`.  So, skipping the first `make distclean`, do
```
make autoconf
./configure
make distclean
```
to update `configure` and have a clean source tree.  Then configure and build Vim as usual.

Also, please note @k-takata 's suggestion iin case `auto/config.mk` is gone for some reason.

**Comment by mcepl** [comment]
> Hm, I am not sure the distclean is needed.

I agree. It [builds](https://copr.fedorainfracloud.org/coprs/build/602973/) even without ``distclean`` and I have refreshed [the patch](https://paste.fedoraproject.org/paste/kwbmjiBIqyL8nuK88UHRxg/).

**Comment by brammool** [comment]
Fixed with 8.0.1111
